### DISCUSSION TITLE
Resampling timeline sample size after every block

### DISCUSSION THREAD

**Discussion by j22tis** [jspsych/jsPsych#1790] [Q&A]
Hi! 

I'm trying to figure out, how to update the value of sample size for a block of trials. I have two kinds of blocks and I want their size (amount of trials in them) to be a randomly sampled number from [11,12,13,14,15] so that it would be resampled after every block.

In my current attempts, I managed to generate a random value for sampler, but it would be the same fixed value throughout the experiment: 
 
# generate amount of trials for block 1, by sampling randomly from 10:15
var size1 = jsPsych.randomization.sampleWithReplacement([10,11,12,13,14,15],1)[0];

# generate amount of trials for block 2 by sampling randomly from 10:15
var size2 = jsPsych.randomization.sampleWithReplacement([10,11,12,13,14,15],1)[0];

# define block 1 
var easy_block = {
     timeline: [experimental_trial],
     timeline_variables: easy_params,
     sample: {
        type: 'with-replacement',
        size: size1, 
        weights: [0.25,0.25,0.75,0.75], 
    }}
# define block 2
var hard_block = {
         timeline: [experimental_trial],
         timeline_variables: hard_params,
         sample: {
            type: 'with-replacement',
            size:  size2, 
            weights: [0.25,0.25,0.75,0.75], 
        }}

# combine blocks into timeline
var blocks = [hard_block, easy_block];
var experimental_timeline = jsPsych.randomization.repeat(blocks, 4)

# add to final procedure
experimental_procedure = {
     timeline: experimental_timeline,
   };

How would I go about resampling the values for size1 and size2 after every block? :)

**Comment by jodeleeuw** [comment]
I don't think you can do this with the `type: 'with-replacement'` method but you could roll your own version by using the `type: 'custom'` method and then you can easily update the variables in between blocks. The function should run anew with each block/repetition.

https://www.jspsych.org/overview/timeline/#custom-sampling-function

**Reply by j22tis** [reply]
Thanks for the pointer! 

I tried something like this:

var easy_params = [
{direction:"right", coherence:"0.4"},
{direction:"left", coherence:"0.4"},
{direction:"right", coherence:"0.8"},
{direction:"left", coherence:"0.8"}]

// hard block parameters (medium trial = 0.4, hard = 0.2 coherence)
var hard_params = [
{direction:"right", coherence:"0.4"},
{direction:"left",coherence:"0.4"},
{direction:"right",coherence:"0.2"},
{direction:"left",coherence:"0.2"}]

Define easy block
var easy_block = {
     timeline: [experimental_trial],
     timeline_variables: easy_params,
     sample: {
type: "custom",
fn: function(t) {
var block_size = jsPsych.randomization.sampleWithReplacement([2,3,4],1)[0];
var t = jsPsych.randomization.sampleWithReplacement(t, block_size, [0.25,0.25,0.75,0.75]);
return t}
}}

Define hard block 
var hard_block = {
         timeline: [experimental_trial],
         timeline_variables: hard_params,
         sample: {
    type: "custom",
    fn: function(t) {
  var block_size = jsPsych.randomization.sampleWithReplacement([2,3,4],1)[0];
  var t = jsPsych.randomization.sampleWithReplacement(t, block_size, [0.25,0.25,0.75,0.75]);
  return t
    }
  }}

Combine to procedure
var experimental_timeline = [hard_block, easy_block];

experimental_procedure = {
     timeline: experimental_timeline,
     repetitions: 4
   };

And it seems to work judgig  by the output data :) Does it seem reasonable?
